docs_src/design/adr/security/0019-spiffe.md

```diff
@@ -0,0 +1,338 @@
+
+### Technical Architecture
+
+![SPIFFE Architecture and Workflow](0019-spiffe-architecture.jpg)
+
+The work flow is as follows:
+
+1. The SPIFFE server is started.
+1. The server creates an initial set of keys.
+1. The trust bundle (certificate authority) data is exported from the SPIFFE server
+   and workload entries are loaded into the SPIFFE server database.
+1. A join token for the SPIFFE agent is created.
+1. The SPIFFE agent is started with the join token to add it to the cluster.
+1. Vault is started and `security-secret-store-setup`
+   initializes it and creates an admin token for `security-spiffe-token-provider` to use.
+1. The `security-spiffe-token-provider` service is started.
+   It obtains an SVID from the SIFFE agent and uses it as a TLS server certificate.
+1. An EdgeX microservice starts and obtains another SVID from the SPIFFE agent
+   and uses it as a TLS client certificate to contact the
+   `security-spiffe-token-provider` service.
+   The EdgeX microservice uses the trust bundle as a server CA
+   to verify the TLS certificate of the remote service.
+1. `security-spiffe-token-provider` verifies the SVID using the trust bundle as client CA
+   to verify the client,
+   extracts the service key,
+   and issues an appropriate Vault service token.
+1. The EdgeX microservice accesses Vault as usual.
+
+
+#### Workload Registration and Agent Sockets
+
+The server uses a workload registration Unix domain socket that allows
+authorization entries to be added to the authorization database.
+This socket is protected by Unix file system permissions to control
+who is allowed to add entries to the database.
+
+In this proposal, a subcommand will be added to the EdgeX `secrets-config`
+utility to simplify the process of registering new services
+that uses the registration socket above.
+
+The agent uses a workload attesation Unix domain socket that
+is open to the world.  This socket is shared via a snap content-interface
+of via a shared host bind mount for Docker.  There is one agent per node.
+
+
+#### Trust Bundle
+
+SVID's must be traceable back to a known issuing authority (certificate authority)
+to determine their validity.
+
+In the proposed implementation, we will generate a CA on first boot and store it persistently.
+This root CA will be distributed as the trust bundle.
+The SPIFFE server will then generate a rotating sub-CA for issuing SVIDs,
+and the issued SVID will include both the leaf certificate and the intermediate certificate.
+
+
+#### Workload Authorization
+
+Workloads are authenticated by connecting to the `spiffe-agent`
+via a Unix domain socket, which is capable of identifying
+the process ID of the remote client.
+The process ID is fed into one of following workload attesters,
+which gather additional metadata about the caller:
+
+* The Unix workload attester gathers UID, GID, path, and SHA-256 hash of the executable.
+  The Unix workload attester would be used for plain services and snaps.
```

Comment:
By plain services, do you mean "native"? Also for snaps, I assume the only selector that would be used would be the path to the executable?

Reply:
Yes, plain == native.  I will update the language.  Yes, only the path would be use as the selector.